File: Completed-Algorithms/LeetCode_20_Valid_Parentheses/valid.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) {
    let stack = []
    let counter = 0
    for (let i = 0; i < s.length; i++) {
        if (s[i] === '(' || s[i]=== '{' || s[i]=== '[') {
            stack.push(s[i])
            counter ++
        } 
        else 
            {
                counter --
                if (s[i] === ')') {
                    if (stack[stack.length - 1] === '(') stack.pop()
                }
                if (s[i] === '}') {
                    if (stack[stack.length - 1] === '{') stack.pop()
                }
                if (s[i] === ']') {
                    if (stack[stack.length - 1] === '[') stack.pop()
                }                
            }

    }
 
    if (stack.length === 0 && counter >= 0) return true
     return false 
  
};

File: Completed-Algorithms/Leetcode20_valid_parenth/valid.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        validParenth = {
            "{": "}",
            "[": "]",
            "(": ")"
        }
        for char in s:
            if char in {"{", "(","["}:
                stack.extend(char)
            elif char in {"}", ")", "]"}:
                if not stack:
                    return False
                last = stack[len(stack) - 1]
                if validParenth[last] != char:
                    return False
                stack.pop()

        return True if len(stack) == 0 else False
